File: plugins/console.cc
#include "../plugin_api.h"
#include <iostream>
#include <sstream>
#include <string>

start();

static void handleCommand(const std::string& cmd) {
    std::istringstream iss(cmd);
    std::string token;
    iss >> token;

    if (token == "help") {
        std::cout << "[console] Commands:\n"
                  << "  load <plugin.dll>\n"
                  << "  unload <plugin.dll>\n"
                  << "  list\n"
                  << "  help\n";
        return;
    }

    if (token == "load") {
        std::string pluginName;
        iss >> pluginName;

        if (pluginName.empty()) {
            plugin::warn("Usage: load <plugin.dll>");
            return;
        }

        plugin::info(std::string("Loading plugin: ").append(pluginName).c_str());
        plugin::host->load_plugin(pluginName.c_str());
        return;
    }

    if (token == "unload") {
        std::string pluginName;
        iss >> pluginName;

        if (pluginName.empty()) {
            plugin::warn("Usage: unload <plugin.dll>");
            return;
        }

        plugin::info(std::string("Unloading plugin: ").append(pluginName).c_str());
        plugin::host->unload_plugin(pluginName.c_str());
        return;
    }

    if (token == "list") {
        plugin::send("requestPluginList", "");
        return;
    }

    plugin::warn(std::string("Unknown command: ").append(token).c_str());
}

event_handler(onConsoleInput) {
    if (!payload) return;
    handleCommand(payload);
}

manifest("console", "1.0.0")

api bool plugin_init(PluginHost* host){
    sethost();
    plugin::on("consoleInput", onConsoleInput);
    return true;
}

api void plugin_shutdown() {
    plugin::off(onConsoleInput);
}

File: plugins/python/console.py
import api
import host
import shlex

@api.on("consoleInput")
def on_console_input(event, payload):
    if not payload:
        return

    try:
        tokens = shlex.split(payload)
    except ValueError:
        api.log("Invalid command syntax", api.WARN)
        return

    if not tokens:
        return

    token = tokens[0]

    if token == "help":
        api.log("[console] Commands:\n"
                "  load <plugin.dll>\n"
                "  unload <plugin.dll>\n"
                "  list\n"
                "  help")
        return

    if token == "load":
        if len(tokens) < 2:
            api.log("Usage: load <plugin.dll>", api.WARN)
            return
        plugin_name = tokens[1]
        api.log(f"Loading plugin: {plugin_name}")
        host.load_plugin(plugin_name)
        return

    if token == "unload":
        if len(tokens) < 2:
            api.log("Usage: unload <plugin.dll>", api.WARN)
            return
        plugin_name = tokens[1]
        api.log(f"Unloading plugin: {plugin_name}")
        host.unload_plugin(plugin_name)
        return

    if token == "list":
        host.send("requestPluginList", "")
        return

    api.log(f"Unknown command: {token}", api.WARN)
